### PR TITLE
Add Acceptance tests and documentation for kubernetes secret resource and data source

### PR DESCRIPTION
**Pull request by Shubbhang351**
1. **What this PR does / why we need it**:
  Add Acceptance tests and documentation for kubernetes secret resource and data source

3. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #[OLYMP-36339](https://jira.eng.vmware.com/browse/OLYMP-36339)

4. **Additional information**

5. **Special notes for your reviewer**

### CODE REVIEW

File: docs/data-sources/cluster.md [hunk 1]
```diff
@@ -322,6 +322,11 @@ Required:
 
 - `aws_availability_zone` (String) The Availability Zone where the AWS nodes are placed
 
+
```
Comment (Vasundharashukla) [inline docs/data-sources/cluster.md:325]: nit: remove these extra lines, they are auto-generated every time, so just check the changes and revert these.

File: docs/data-sources/kubernetes_secret.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
+---
+Title: "Secret Resource"
+Description: |-
+    Fetching the details of sceret resource.
+---
+
+# Namespace
+
+Manage a sceret using this Terraform module.
```
Comment (Vasundharashukla) [inline docs/data-sources/kubernetes_secret.md:9]: Please get this content reviewed by the feature owners. They can suggest better what needs to be added. Also @ishangupta-ds can take a look at this.

File: docs/data-sources/kubernetes_secret.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
+---
+Title: "Secret Resource"
+Description: |-
+    Fetching the details of sceret resource.
+---
+
+# Namespace
```
Comment (Vasundharashukla) [inline docs/data-sources/kubernetes_secret.md:7]: Please correct the title- Secret/ Kubernetes Secret

File: docs/resources/cluster.md [hunk 4]
```diff
@@ -1029,6 +1029,11 @@ Optional:
 - `disk_size` (String) Root disk size in gigabytes for the VM
 - `memory` (String) Memory associated with the node in megabytes
 
+
```
Comment (Vasundharashukla) [inline docs/resources/cluster.md:1032]: nit: remove these extra lines, they are auto-generated every time, so just check the changes and revert these.

File: docs/resources/custom_policy.md [hunk 5]
```diff
@@ -1105,6 +1105,7 @@ Optional:
 - `cluster` (Block List, Max: 1) The schema for cluster policy full name (see [below for nested schema](#nestedblock--scope--cluster))
 - `cluster_group` (Block List, Max: 1) The schema for cluster group policy full name (see [below for nested schema](#nestedblock--scope--cluster_group))
 - `organization` (Block List, Max: 1) The schema for organization policy full name (see [below for nested schema](#nestedblock--scope--organization))
+- `workspace` (Block List, Max: 1) The schema for workspace policy full name (see [below for nested schema](#nestedblock--scope--workspace))
```
Comment (Vasundharashukla) [inline docs/resources/custom_policy.md:1108]: Revert this change

File: docs/resources/custom_policy.md [hunk 6]
```diff
@@ -1135,6 +1136,14 @@ Required:
 - `organization` (String) ID of this organization
 
 
+<a id="nestedblock--scope--workspace"></a>
```
Comment (Vasundharashukla) [inline docs/resources/custom_policy.md:1139]: Revert this change

File: docs/resources/image_policy.md [hunk 7]
```diff
@@ -455,9 +455,31 @@ resource "tanzu-mission-control_image_policy" "organization_scoped_require-diges
 
 Optional:
 
+- `cluster` (Block List, Max: 1) The schema for cluster policy full name (see [below for nested schema](#nestedblock--scope--cluster))
```
Comment (Vasundharashukla) [inline docs/resources/image_policy.md:458]: Revert this change

File: docs/resources/image_policy.md [hunk 8]
```diff
@@ -455,9 +455,31 @@ resource "tanzu-mission-control_image_policy" "organization_scoped_require-diges
...
+- `cluster_group` (Block List, Max: 1) The schema for cluster group policy full name (see [below for nested schema](#nestedblock--scope--cluster_group))
 - `organization` (Block List, Max: 1) The schema for organization policy full name (see [below for nested schema](#nestedblock--scope--organization))
 - `workspace` (Block List, Max: 1) The schema for workspace policy full name (see [below for nested schema](#nestedblock--scope--workspace))
 
+<a id="nestedblock--scope--cluster"></a>
```
Comment (Vasundharashukla) [inline docs/resources/image_policy.md:463]: Revert this change

File: docs/resources/namespace_quota_policy.md [hunk 9]
```diff
@@ -639,6 +639,7 @@ Optional:
 - `cluster` (Block List, Max: 1) The schema for cluster policy full name (see [below for nested schema](#nestedblock--scope--cluster))
 - `cluster_group` (Block List, Max: 1) The schema for cluster group policy full name (see [below for nested schema](#nestedblock--scope--cluster_group))
 - `organization` (Block List, Max: 1) The schema for organization policy full name (see [below for nested schema](#nestedblock--scope--organization))
+- `workspace` (Block List, Max: 1) The schema for workspace policy full name (see [below for nested schema](#nestedblock--scope--workspace))
```
Comment (Vasundharashukla) [inline docs/resources/namespace_quota_policy.md:642]: Revert this change

File: docs/resources/namespace_quota_policy.md [hunk 10]
```diff
@@ -669,6 +670,15 @@ Required:
 - `organization` (String) ID of this organization
 
 
+<a id="nestedblock--scope--workspace"></a>
```
Comment (Vasundharashukla) [inline docs/resources/namespace_quota_policy.md:673]: Revert this change

File: docs/resources/namespace_quota_policy.md [hunk 11]
```diff
@@ -686,9 +696,9 @@ Optional:
 Optional:
 
 - `custom` (Block List, Max: 1) The input schema for namespace quota policy custom recipe version v1 (see [below for nested schema](#nestedblock--spec--input--custom))
-- `large` (Block List, Max: 1) The input schema for namespace quota policy large recipe version v1 
-- `medium` (Block List, Max: 1) The input schema for namespace quota policy medium recipe version v1 
-- `small` (Block List, Max: 1) The input schema for namespace quota policy small recipe version v1 
+- `large` (Block List, Max: 1) The input schema for namespace quota policy large recipe version v1 (see [below for nested schema](#nestedblock--spec--input--large))
```
Comment (Vasundharashukla) [inline docs/resources/namespace_quota_policy.md:699]: Revert this change

File: docs/resources/namespace_quota_policy.md [hunk 12]
```diff
@@ -706,6 +716,19 @@ Optional:
 - `resource_counts` (Map of Number) The total number of Services of the given type that can exist in a namespace
 
 
+<a id="nestedblock--spec--input--large"></a>
```
Comment (Vasundharashukla) [inline docs/resources/namespace_quota_policy.md:719]: Revert this change

File: docs/resources/security_policy.md [hunk 13]
```diff
@@ -776,6 +776,7 @@ Optional:
 - `cluster` (Block List, Max: 1) The schema for cluster policy full name (see [below for nested schema](#nestedblock--scope--cluster))
 - `cluster_group` (Block List, Max: 1) The schema for cluster group policy full name (see [below for nested schema](#nestedblock--scope--cluster_group))
 - `organization` (Block List, Max: 1) The schema for organization policy full name (see [below for nested schema](#nestedblock--scope--organization))
+- `workspace` (Block List, Max: 1) The schema for workspace policy full name (see [below for nested schema](#nestedblock--scope--workspace))
```
Comment (Vasundharashukla) [inline docs/resources/security_policy.md:779]: Revert this change

File: docs/resources/security_policy.md [hunk 14]
```diff
@@ -806,6 +807,15 @@ Required:
 - `organization` (String) ID of this organization
 
 
+<a id="nestedblock--scope--workspace"></a>
```
Comment (Vasundharashukla) [inline docs/resources/security_policy.md:810]: Revert this change

File: internal/resources/kubernetessecret/resource_secret.go [hunk 15]
```diff
@@ -49,15 +49,16 @@ var secretSchema = map[string]*schema.Schema{
 		ForceNew:    true,
 	},
 	OrgIdKey: {
-		Type: schema.TypeString,
+		Type:     schema.TypeString,
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:52]: Description to be added here

File: templates/data-sources/kubernetes_secret.md.tmpl [hunk 16]
```diff
@@ -0,0 +1,23 @@
+---
+Title: "Secret Resource"
+Description: |-
+    Fetching the details of sceret resource.
+---
+
+# Namespace
+
+Manage a sceret using this Terraform module.
```
Comment (Vasundharashukla) [inline templates/data-sources/kubernetes_secret.md.tmpl:9]: Should there be some content difference between resource and data source, may be refer to the other docs and get it reviewed by @ishangupta-ds and the resource owners.

**Review by Vasundharashukla — CHANGES_REQUESTED** [review]

File: docs/data-sources/cluster.md (hunk 1 above)
Comment (Shubbhang351) [reply docs/data-sources/cluster.md]: revert this changes

File: docs/resources/security_policy.md (hunk 14 above)
Comment (Shubbhang351) [reply docs/resources/security_policy.md]: reverted these changes

File: templates/data-sources/kubernetes_secret.md.tmpl (hunk 16 above)
Comment (Shubbhang351) [reply templates/data-sources/kubernetes_secret.md.tmpl]: @ishangupta-ds PTAL, at docs

File: docs/data-sources/kubernetes_secret.md (hunk 3 above)
Comment (Shubbhang351) [reply docs/data-sources/kubernetes_secret.md]: updated PTAL

File: docs/data-sources/kubernetes_secret.md (hunk 2 above)
Comment (Shubbhang351) [reply docs/data-sources/kubernetes_secret.md]: @ishangupta-ds PTAL

**Comment by Shubbhang351** [comment]
Screenshot 2023-03-30 at 1 19 02 PM


output for Acceptance tests

File: docs/resources/kubernetes_secret.md [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+# Kubernetes Secret
+
+Manage a Kubernetes Secret using this Terraform module.
+
+In managed clusters, both attached and provisioned, you can create Kubernetes Secret that you can manage through Tanzu Mission Control.
+
+To create a cluster secret, you must be associated with the cluster.admin role
+
+[cluster Secret]: https://docs.vmware.com/en/VMware-Tanzu-Mission-Control/services/tanzumc-using/GUID-BBE2404D-C2EE-41C7-B639-C0322783A74D.html
+
```
Comment (ishangupta-ds) [inline docs/resources/kubernetes_secret.md:16]: Can you please also add a few lines about secret export and how that works?

File: docs/resources/kubernetes_secret.md [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+# Kubernetes Secret
+
+Manage a Kubernetes Secret using this Terraform module.
+
+In managed clusters, both attached and provisioned, you can create Kubernetes Secret that you can manage through Tanzu Mission Control.
+
+To create a cluster secret, you must be associated with the cluster.admin role
+
+[cluster Secret]: https://docs.vmware.com/en/VMware-Tanzu-Mission-Control/services/tanzumc-using/GUID-BBE2404D-C2EE-41C7-B639-C0322783A74D.html
```
Comment (ishangupta-ds) [inline docs/resources/kubernetes_secret.md:15]: Please add another reference for secret export from the TMC documentation.

File: docs/resources/kubernetes_secret.md [hunk 19]
```diff
@@ -0,0 +1,121 @@
+---
+Title: "Kubernetes Secret Resource"
+Description: |-
+    Creating the Kubernetes Secret resource.
+---
+
+# Kubernetes Secret
```
Comment (ishangupta-ds) [inline docs/resources/kubernetes_secret.md:7]: Please add a section for scope, refer other resources like - `security policy`

**Review by ishangupta-ds — CHANGES_REQUESTED** [review]

File: docs/resources/kubernetes_secret.md (hunk 19 above)
Comment (Shubbhang351) [reply docs/resources/kubernetes_secret.md:7]: added PTAL

File: docs/resources/kubernetes_secret.md (hunk 18 above)
Comment (Shubbhang351) [reply docs/resources/kubernetes_secret.md:19]: added PTAL

File: docs/resources/kubernetes_secret.md [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+- `name` (String) Name of the secret resource.
+- `namespace_name` (String) Name of Namespace where secret will be created.
+- `scope` (Block List, Min: 1, Max: 1) Scope for the secret having one of the valid scopes for secret: currently we have only cluster scope (see [below for nested schema](#nestedblock--scope))
+
+### Optional
+
+- `export` (Boolean) Export the secret to all namespaces.
+- `meta` (Block List, Max: 1) Metadata for the resource (see [below for nested schema](#nestedblock--meta))
+- `org_id` (String) ID of Organization.
+- `spec` (Block List, Max: 1) Spec for the kubernetes secret (see [below for nested schema](#nestedblock--spec))
```
Comment (narahari92) [inline docs/resources/kubernetes_secret.md:81]: Is Spec optional? Shouldn't it be required?

Comment (Shubbhang351) [reply docs/resources/kubernetes_secret.md:81]: for data source[Get request for cluster-secret] and resource[Create, delete, update for cluster-secret] we have used the same schema and for data source we don't need spec part to mention, only for resource we need spec so that's why I have made it to Optional.

@ishangupta-ds , can you also take a look

File: docs/data-sources/kubernetes_secret.md (hunk 2 above)
Comment (ishangupta-ds) [reply docs/data-sources/kubernetes_secret.md]: Left a few comments.

File: docs/resources/kubernetes_secret.md (hunk 20 above)
Comment (ishangupta-ds) [reply docs/resources/kubernetes_secret.md:81]: @shreyassreenivas do we follow the same for other resources which have corresponding data sources? If yes then can we keep the spec option for the `kubernetes_secret` resource as well? Cc: @Vasundharashukla @narahari92

**Comment by ishangupta-ds** [comment]
@Shubbhang351 Can we attach the screenshots of the acceptance tests?

**Comment by Shubbhang351** [comment]
Screenshot 2023-03-30 at 1 19 02 PM

output for the acceptance tests

**Comment by shreyassreenivas** [comment]
@Shubbhang351 , There are two types acceptance tests needed, one what you have implemented and other is acceptance tests with mocks

File: docs/resources/kubernetes_secret.md (hunk 20 above)
Comment (Shubbhang351) [reply docs/resources/kubernetes_secret.md]: @narahari92 Please take a look, we have fixed the docs and Schema 
**spec** is a required block for Resource and Read only for Data Source.

cc: @ishangupta-ds, @Vasundharashukla , @shreyassreenivas

File: internal/resources/kubernetessecret/resource_secret.go [hunk 21]
```diff
@@ -34,52 +34,67 @@ func ResourceSecret() *schema.Resource {
...
-		Optional:    true,
-		Default:     false,
-	},
-	statusKey: {
-		Type:        schema.TypeMap,
-		Description: "Status for the Secret Export.",
-		Computed:    true,
-		Elem:        &schema.Schema{Type: schema.TypeString},
-	},
+func getSecretSchema(isDataSource bool) map[string]*schema.Schema {
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/resource_secret.go:41]: This function is not intuitive. 
So, can you add functional pararement 
getSecretSchema(forDatasource), getSecretSchema(forResource)
Implement something in these lines 

or two new functions which will call this function

Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:49]: Added two new function **getDataSourceSchema()** , **getResourceSchema()** which are calling **getSecretSchema(isDataSource bool)**

Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/resource_secret.go:49]: Also add unit tests and ensure that the schema does not change

File: internal/resources/kubernetessecret/resource_secret.go [hunk 22]
```diff
@@ -89,9 +104,11 @@ type dockerConfigJSON struct {
 var secretSpec = &schema.Schema{
 	Type:        schema.TypeList,
 	Description: "Spec for the kubernetes secret",
-	Optional:    true,
-	MaxItems:    1,
-	MinItems:    1,
+	Required:    true,
+	// Optional:    false,
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: removed

File: internal/resources/kubernetessecret/resource_secret.go (hunk 21 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:49]: added PTAL

Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:49]: output

```
=== RUN   TestGetSecretSchema
=== RUN   TestGetSecretSchema/resource_schema_test
--- PASS: TestGetSecretSchema/resource_schema_test (0.00s)
=== RUN   TestGetSecretSchema/data_source_schema_test
--- PASS: TestGetSecretSchema/data_source_schema_test (0.00s)
=== RUN   TestGetSecretSchema/generate_wrong_resource_schema
--- PASS: TestGetSecretSchema/generate_wrong_resource_schema (0.00s)
--- PASS: TestGetSecretSchema (0.00s)
PASS
ok      github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/kubernetessecret  1.782s


> Test run finished at 12/04/2023, 18:49:23 <

```

File: docs/resources/kubernetes_secret.md (hunk 20 above)
Comment (narahari92) [reply docs/resources/kubernetes_secret.md]: LGTM

Comment (ishangupta-ds) [reply docs/resources/kubernetes_secret.md]: Resolving as per the discussion with @narahari92

**Review by ishangupta-ds — APPROVED** [review]
Please address the comments from @shreyassreenivas. The rest looks good to me!

File: internal/resources/kubernetessecret/resource_secret.go (hunk 21 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:49]: PTAL

**Comment by Shubbhang351** [comment]
@ishangupta-ds this jira I've created https://jira.eng.vmware.com/browse/OLYMP-38047
for mocks acceptance tests

**Review by shreyassreenivas — APPROVED** [review]
